src/rust/engine/src/graph.rs

```diff
@@ -443,6 +445,36 @@ impl InnerGraph {
     try!(f.write_all(b"\n"));
     Ok(())
   }
+
+  pub fn all_digests(&self, core: Arc<Core>) -> Vec<hashing::Digest> {
+    self
+      .nodes
+      .keys()
+      .map(|key| self.entry(key))
+      .filter_map(|maybe_entry| maybe_entry)
+      .map(|entry| entry.node.clone())
+      .filter_map(|node| {
+        let node_key = node.content();
+        match (node_key, self.entry_id(&node)) {
+          (&NodeKey::DigestFile(ref digest_file), Some(entry_id)) => Some((
+            entry_id.clone(),
+            digest_file.clone(),
+          )),
+          _ => None,
+        }
+      })
+      .filter_map(|(entry_id, digest_file)| {
+        let context = Context {
+          entry_id: entry_id,
+          core: core.clone(),
+        };
+        match digest_file.run(context).poll() {
```

Comment:
You should not need to recreate the node to poll it. Given an `Entry`, you'd call `entry.peek()` to get the current state if it exists (and to not create it if it doesn't).

This allows for a race where if something is running, its lease won't be extended. But IMO that's not worth waiting for it to complete.

Reply:
Done :)